File: gooseberry/packages/server/src/projects/projects.ts
import { randomUUID } from "node:crypto";
import { basename, isAbsolute, relative } from "node:path";
import { normalizeProjectIcon, normalizeProjectName, type Project } from "@gooseberry/contracts";
import { canonicalPath } from "../git";
import { assertMountedDirectory, assertMountedProject } from "../path-admission";
import { loadProjects, saveProjects } from "../persistence";

type ProjectPublisher = (project: Project) => void;

let publishProject: ProjectPublisher | null = null;

export function setProjectPublisher(fn: ProjectPublisher | null): void {
	publishProject = fn;
}

function emit(project: Project): void {
	publishProject?.(project);
}

function slugify(name: string): string {
	return (
		name
			.toLowerCase()
			.replace(/[^a-z0-9]+/g, "-")
			.replace(/^-+|-+$/g, "") || "project"
	);
}

function uniqueSlug(base: string, taken: Set<string>): string {
	if (!taken.has(base)) return base;
	let n = 2;
	while (taken.has(`${base}-${n}`)) n += 1;
	return `${base}-${n}`;
}

function ensureSlugs(projects: Project[]): boolean {
	const taken = new Set(projects.map((p) => p.slug).filter(Boolean));
	let changed = false;
	for (const project of projects) {
		if (!project.slug) {
			project.slug = uniqueSlug(slugify(project.name), taken);
			taken.add(project.slug);
			changed = true;
		}
	}
	return changed;
}

export function getProjects(): Project[] {
	const projects = loadProjects();
	for (const project of projects) {
		if (project.roots.length === 0) throw new Error(`Project has no roots: ${project.id}`);
		project.roots = project.roots.map((root) => assertMountedProject(root));
	}
	if (ensureSlugs(projects)) saveProjects(projects);
	return projects;
}

export function openProject(path: string): Project {
	const mountedRoot = assertMountedProject(path);

	const projects = getProjects();
	const existing = projects.find((project) =>
		project.roots.some((root) => canonicalPath(root) === mountedRoot),
	);
	if (existing) {
		delete existing.closed;
		existing.lastOpened = Date.now();
		saveProjects(projects);
		emit(existing);
		return existing;
	}

	const taken = new Set(projects.map((p) => p.slug));
	const project: Project = {
		id: randomUUID(),
		name: basename(mountedRoot),
		roots: [mountedRoot],
		slug: uniqueSlug(slugify(basename(mountedRoot)), taken),
		lastOpened: Date.now(),
	};
	projects.push(project);
	saveProjects(projects);
	emit(project);
	return project;
}

export function getProject(id: string): Project {
	const project = getProjects().find((candidate) => candidate.id === id);
	if (!project) throw new Error(`Unknown project: ${id}`);
	return project;
}

export function addProjectRoot(id: string, path: string): Project {
	const root = assertMountedProject(path);
	const projects = getProjects();
	const project = projects.find((candidate) => candidate.id === id);
	if (!project) throw new Error(`Unknown project: ${id}`);
	const owner = projects.find(
		(candidate) =>
			candidate.id !== id && candidate.roots.some((known) => canonicalPath(known) === root),
	);
	if (owner) throw new Error(`Directory is already a root of project ${owner.name}`);
	if (!project.roots.some((known) => canonicalPath(known) === root)) project.roots.push(root);
	project.lastOpened = Date.now();
	saveProjects(projects);
	emit(project);
	return project;
}

export function removeProjectRoot(id: string, path: string): Project {
	const wanted = canonicalPath(assertMountedProject(path));
	const projects = getProjects();
	const project = projects.find((candidate) => candidate.id === id);
	if (!project) throw new Error(`Unknown project: ${id}`);
	if (project.roots.length === 1) throw new Error("A project must keep at least one root");
	const next = project.roots.filter((root) => canonicalPath(root) !== wanted);
	if (next.length === project.roots.length) throw new Error("Project root not found");
	project.roots = next;
	saveProjects(projects);
	emit(project);
	return project;
}

export function updateProject(id: string, update: { name?: unknown; icon?: unknown }): Project {
	if (update.name === undefined && update.icon === undefined) {
		throw new Error("Project update requires a name or icon");
	}
	const projects = getProjects();
	const project = projects.find((candidate) => candidate.id === id);
	if (!project) throw new Error(`Unknown project: ${id}`);
	if (update.name !== undefined) project.name = normalizeProjectName(update.name);
	if (update.icon !== undefined) project.icon = normalizeProjectIcon(update.icon);
	project.lastOpened = Date.now();
	saveProjects(projects);
	emit(project);
	return project;
}

function isWithin(root: string, candidate: string): boolean {
	const rel = relative(root, candidate);
	return rel === "" || (!rel.startsWith("..") && !isAbsolute(rel));
}

export function assertProjectCwd(projectId: string, cwd?: string): string {
	const project = getProject(projectId);
	const defaultRoot = project.roots[0];
	if (!defaultRoot) throw new Error(`Project has no roots: ${projectId}`);
	const candidate = assertMountedDirectory(cwd?.trim() || defaultRoot, "Session directory");
	if (!project.roots.some((root) => isWithin(root, candidate))) {
		throw new Error("Session directory is outside the project roots");
	}
	return candidate;
}

function newestFirst(projects: Project[]): Project[] {
	return projects.sort((a, b) => b.lastOpened - a.lastOpened);
}

export function listProjects(): Project[] {
	return newestFirst(getProjects().filter((project) => project.closed !== true));
}

export function listRecentProjects(): Project[] {
	return newestFirst(getProjects());
}

export function closeProject(id: string): Project {
	const projects = getProjects();
	const project = projects.find((candidate) => candidate.id === id);
	if (!project) throw new Error(`Unknown project: ${id}`);
	project.closed = true;
	saveProjects(projects);
	emit(project);
	return project;
}

File: gooseberry/internal/workspace/projects.go
package workspace

import (
	"cmp"
	"fmt"
	"path/filepath"
	"regexp"
	"slices"
	"strings"
	"sync"
	"time"
	"unicode/utf16"

	"github.com/miloszkolber/gooseberry/internal/identifier"
	"github.com/miloszkolber/gooseberry/internal/persist"
)

var (
	slugCharacters = regexp.MustCompile(`[^a-z0-9]+`)
	projectIcons   = map[string]bool{"folder": true, "code": true, "book": true, "flask": true, "rocket": true, "sparkles": true}
)

type Projects struct {
	mu       sync.RWMutex
	store    persist.Store
	policy   *PathPolicy
	now      func() time.Time
	publish  func(Project)
	loaded   bool
	projects []Project
}

func NewProjects(store persist.Store, policy *PathPolicy) *Projects {
	return &Projects{store: store, policy: policy, now: time.Now}
}

func utf16Length(value string) int {
	count := 0
	for _, character := range value {
		count += utf16.RuneLen(character)
	}
	return count
}

func (p *Projects) SetPublisher(publish func(Project)) {
	p.mu.Lock()
	p.publish = publish
	p.mu.Unlock()
}

type persistedProject struct {
	Project
	Path string `json:"path,omitempty"`
}

func (p *Projects) load() ([]Project, error) {
	// This process is the sole owner of project state. Load and validate once,
	// then keep a private snapshot synchronized with each atomic write.
	if p.loaded {
		return cloneProjects(p.projects), nil
	}
	var persisted []persistedProject
	for _, name := range []string{"projects.json", "projects.json.bak"} {
		raw, _, err := persist.ReadFile(filepath.Join(p.store.Dir, name))
		if err != nil {
			continue
		}
		var decoded []persistedProject
		if persist.Decode(raw, &decoded, validateProjects) != nil {
			continue
		}
		persisted = decoded
		break
	}
	projects := make([]Project, 0, len(persisted))
	migrated := false
	for _, entry := range persisted {
		roots := slices.Clone(entry.Roots)
		if len(roots) == 0 && entry.Path != "" {
			roots = []string{entry.Path}
			migrated = true
		}
		if entry.ID == "" || len(roots) == 0 {
			continue
		}
		for index, root := range roots {
			canonical, resolveErr := p.policy.Directory(root, "Project")
			if resolveErr != nil {
				return nil, resolveErr
			}
			roots[index] = canonical
		}
		projects = append(projects, Project{
			ID: entry.ID, Name: entry.Name, Roots: roots, Slug: entry.Slug,
			LastOpened: entry.LastOpened, Icon: entry.Icon, Closed: entry.Closed,
		})
	}
	if ensureSlugs(projects) {
		migrated = true
	}
	if migrated {
		if err := p.save(projects); err != nil {
			return nil, err
		}
	} else {
		p.projects = cloneProjects(projects)
		p.loaded = true
	}
	return cloneProjects(projects), nil
}

func (p *Projects) ensureLoaded() error {
	p.mu.RLock()
	if p.loaded {
		p.mu.RUnlock()
		return nil
	}
	p.mu.RUnlock()

	p.mu.Lock()
	defer p.mu.Unlock()
	_, err := p.load()
	return err
}

func (p *Projects) save(projects []Project) error {
	values := make([]persistedProject, len(projects))
	for index, project := range projects {
		values[index].Project = project
	}
	if err := persist.Write(p.store, "projects.json", values, validateProjects); err != nil {
		return err
	}
	p.projects = cloneProjects(projects)
	p.loaded = true
	return nil
}

func cloneProjects(projects []Project) []Project {
	result := make([]Project, len(projects))
	for index, project := range projects {
		result[index] = project
		result[index].Roots = slices.Clone(project.Roots)
	}
	return result
}

func validateProjects(values []persistedProject) error {
	if values == nil {
		return fmt.Errorf("projects must be an array")
	}
	for _, value := range values {
		if value.ID == "" || value.Roots == nil && value.Path == "" || value.Icon != "" && !projectIcons[value.Icon] {
			return fmt.Errorf("invalid persisted project")
		}
	}
	return nil
}

func (p *Projects) List(includeClosed bool) ([]Project, error) {
	if err := p.ensureLoaded(); err != nil {
		return nil, err
	}
	p.mu.RLock()
	defer p.mu.RUnlock()
	filtered := make([]Project, 0, len(p.projects))
	for _, project := range p.projects {
		if includeClosed || !project.Closed {
			project.Roots = slices.Clone(project.Roots)
			filtered = append(filtered, project)
		}
	}
	slices.SortStableFunc(filtered, func(a, b Project) int { return cmp.Compare(b.LastOpened, a.LastOpened) })
	return filtered, nil
}

func (p *Projects) Get(id string) (Project, error) {
	if err := p.ensureLoaded(); err != nil {
		return Project{}, err
	}
	p.mu.RLock()
	defer p.mu.RUnlock()
	var result Project
	found := false
	for _, project := range p.projects {
		if project.ID == id && (!found || project.LastOpened > result.LastOpened) {
			result, found = project, true
		}
	}
	if !found {
		return Project{}, fmt.Errorf("unknown project: %s", id)
	}
	result.Roots = slices.Clone(result.Roots)
	return result, nil
}

func (p *Projects) Open(path string) (result Project, err error) {
	p.mu.Lock()
	defer p.finishMutation(&result, &err)
	root, err := p.policy.Directory(path, "Project")
	if err != nil {
		return Project{}, err
	}
	projects, err := p.load()
	if err != nil {
		return Project{}, err
	}
	for index := range projects {
		for _, known := range projects[index].Roots {
			if known == root {
				projects[index].Closed = false
				projects[index].LastOpened = float64(p.now().UnixMilli())
				if err := p.save(projects); err != nil {
					return Project{}, err
				}
				return projects[index], nil
			}
		}
	}
	name := filepath.Base(root)
	taken := make(map[string]bool, len(projects))
	for _, project := range projects {
		taken[project.Slug] = true
	}
	project := Project{ID: identifier.New(), Name: name, Roots: []string{root}, Slug: uniqueSlug(slugify(name), taken), LastOpened: float64(p.now().UnixMilli())}
	projects = append(projects, project)
	if err := p.save(projects); err != nil {
		return Project{}, err
	}
	return project, nil
}

func (p *Projects) AddRoot(id, path string) (result Project, err error) {
	p.mu.Lock()
	defer p.finishMutation(&result, &err)
	root, err := p.policy.Directory(path, "Project")
	if err != nil {
		return Project{}, err
	}
	projects, err := p.load()
	if err != nil {
		return Project{}, err
	}
	index := projectIndex(projects, id)
	if index < 0 {
		return Project{}, fmt.Errorf("unknown project: %s", id)
	}
	for otherIndex, other := range projects {
		if otherIndex != index && contains(other.Roots, root) {
			return Project{}, fmt.Errorf("directory is already a root of project %s", other.Name)
		}
	}
	if !contains(projects[index].Roots, root) {
		projects[index].Roots = append(projects[index].Roots, root)
	}
	projects[index].LastOpened = float64(p.now().UnixMilli())
	if err := p.save(projects); err != nil {
		return Project{}, err
	}
	return projects[index], nil
}

func (p *Projects) RemoveRoot(id, path string) (result Project, err error) {
	p.mu.Lock()
	defer p.finishMutation(&result, &err)
	root, err := p.policy.Directory(path, "Project")
	if err != nil {
		return Project{}, err
	}
	projects, err := p.load()
	if err != nil {
		return Project{}, err
	}
	index := projectIndex(projects, id)
	if index < 0 {
		return Project{}, fmt.Errorf("unknown project: %s", id)
	}
	if len(projects[index].Roots) == 1 {
		return Project{}, fmt.Errorf("a project must keep at least one root")
	}
	next := make([]string, 0, len(projects[index].Roots)-1)
	for _, known := range projects[index].Roots {
		if known != root {
			next = append(next, known)
		}
	}
	if len(next) == len(projects[index].Roots) {
		return Project{}, fmt.Errorf("project root not found")
	}
	projects[index].Roots = next
	if err := p.save(projects); err != nil {
		return Project{}, err
	}
	return projects[index], nil
}

func (p *Projects) Update(id string, name, icon *string) (result Project, err error) {
	if name == nil && icon == nil {
		return Project{}, fmt.Errorf("project update requires a name or icon")
	}
	p.mu.Lock()
	defer p.finishMutation(&result, &err)
	projects, err := p.load()
	if err != nil {
		return Project{}, err
	}
	index := projectIndex(projects, id)
	if index < 0 {
		return Project{}, fmt.Errorf("unknown project: %s", id)
	}
	if name != nil {
		normalized := strings.TrimSpace(*name)
		if normalized == "" || strings.ContainsRune(normalized, 0) || utf16Length(normalized) > 100 {
			return Project{}, fmt.Errorf("invalid project name")
		}
		projects[index].Name = normalized
	}
	if icon != nil {
		if !projectIcons[*icon] {
			return Project{}, fmt.Errorf("unknown project icon")
		}
		projects[index].Icon = *icon
	}
	projects[index].LastOpened = float64(p.now().UnixMilli())
	if err := p.save(projects); err != nil {
		return Project{}, err
	}
	return projects[index], nil
}

func (p *Projects) Close(id string) (result Project, err error) {
	p.mu.Lock()
	defer p.finishMutation(&result, &err)
	projects, err := p.load()
	if err != nil {
		return Project{}, err
	}
	index := projectIndex(projects, id)
	if index < 0 {
		return Project{}, fmt.Errorf("unknown project: %s", id)
	}
	projects[index].Closed = true
	if err := p.save(projects); err != nil {
		return Project{}, err
	}
	return projects[index], nil
}

func (p *Projects) finishMutation(result *Project, err *error) {
	publish := p.publish
	p.mu.Unlock()
	if *err == nil && publish != nil {
		publish(*result)
	}
}

func (p *Projects) AssertCWD(projectID, cwd string) (string, error) {
	project, err := p.Get(projectID)
	if err != nil {
		return "", err
	}
	if strings.TrimSpace(cwd) == "" {
		cwd = project.Roots[0]
	}
	candidate, err := p.policy.Directory(cwd, "Session directory")
	if err != nil {
		return "", err
	}
	for _, root := range project.Roots {
		if Within(root, candidate) {
			return candidate, nil
		}
	}
	return "", fmt.Errorf("session directory is outside the project roots")
}

func (p *Projects) AssertRoot(projectID, root string) (string, error) {
	project, err := p.Get(projectID)
	if err != nil {
		return "", err
	}
	// Get already resolves and authorizes each root against the current mount.
	if contains(project.Roots, root) {
		return root, nil
	}
	candidate, err := p.policy.Directory(root, "Project root")
	if err != nil {
		return "", err
	}
	if !contains(project.Roots, candidate) {
		return "", fmt.Errorf("project root must exactly match an admitted project root")
	}
	return candidate, nil
}

func projectIndex(projects []Project, id string) int {
	for index := range projects {
		if projects[index].ID == id {
			return index
		}
	}
	return -1
}

func slugify(name string) string {
	slug := strings.Trim(slugCharacters.ReplaceAllString(strings.ToLower(name), "-"), "-")
	if slug == "" {
		return "project"
	}
	return slug
}

func uniqueSlug(base string, taken map[string]bool) string {
	if !taken[base] {
		return base
	}
	for suffix := 2; ; suffix++ {
		candidate := fmt.Sprintf("%s-%d", base, suffix)
		if !taken[candidate] {
			return candidate
		}
	}
}

func ensureSlugs(projects []Project) bool {
	if !slices.ContainsFunc(projects, func(project Project) bool { return project.Slug == "" }) {
		return false
	}
	taken := make(map[string]bool, len(projects))
	for _, project := range projects {
		if project.Slug != "" {
			taken[project.Slug] = true
		}
	}
	changed := false
	for index := range projects {
		if projects[index].Slug == "" {
			projects[index].Slug = uniqueSlug(slugify(projects[index].Name), taken)
			taken[projects[index].Slug] = true
			changed = true
		}
	}
	return changed
}

func contains(values []string, wanted string) bool {
	for _, value := range values {
		if value == wanted {
			return true
		}
	}
	return false
}
